code-dot-org/code-dot-org
# Add HoC markdown files to the translation pipeline

[FND-1329](https://codedotorg.atlassian.net/browse/FND-1329): Make code.org/hourofcode/* pages translatable.

Before adding HoC markdown files to the translation pipeline, we have to remove all Ruby code and iframe from them. This means:
1. For `md.erb` file, remove its Ruby code and change the file extension to just `md`.
2. For all `md` files that use iframe to display Youtube videos, convert them to `md.partial` and use the `display_video_thumbnail` partial instead. (Additionally, always center the videos and remove the black bars above and below them.)

Then we can finally add a list of safe markdown files to `sync-in.rb`.

## Screenshots


| Before | After |
|--|--|
|<img width="650" alt="Screen Shot 2020-11-05 at 11 00 51 AM" src="https://user-images.githubusercontent.com/46507039/98196258-88b0f600-1ed8-11eb-8709-503340bada89.png">|<img width="647" alt="Screen Shot 2020-11-05 at 11 01 02 AM" src="https://user-images.githubusercontent.com/46507039/98196262-8b135000-1ed8-11eb-894d-e6365fe579e4.png">|
|<img width="633" alt="Screen Shot 2020-11-05 at 11 03 20 AM" src="https://user-images.githubusercontent.com/46507039/98196384-d4fc3600-1ed8-11eb-8daf-04fbffb3abd4.png">|<img width="624" alt="Screen Shot 2020-11-05 at 11 03 25 AM" src="https://user-images.githubusercontent.com/46507039/98196393-d9c0ea00-1ed8-11eb-8819-1dd2affbc375.png">|

## Links
- [Slack thread](https://codedotorg.slack.com/archives/CFTFD6BPV/p1604518100304800) on making a markdown file _safe_ vs. _translator_friendly_.

## Testing story
Steps to verify that `pegasus/sites.v3/code.org/public/hourofcode/starwars.md` is translatable:

1. Run the `localize_markdown_content` function in `bin/i18n/sync-in.rb`. Verify that the `starwars.md.partial` file is copied to `i18n/locales/source/markdown/public/hourofcode/starwars.md`.
2. Simulate the result of `sync-down.rb` step by copying a translated version of `starwars.md` to `i18n/locales/<locale>/codeorg-markdown/hourofcode/starwars.md`. An example locale is `vi-VN`.
3. Run `bin/i18n/sync-out.rb`, specifically the `distribute_translations` method, and verify that the translated file is copied to `pegasus/sites.v3/code.org/i18n/public/hourofcode/starwars.vi-VN.md.partial`. 
4. Open `http://localhost.code.org:3000/lang/vi/hourofcode/starwars`. The page should be translated now and the Starwars intro video still shows up correctly.
